sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala

```diff
@@ -0,0 +1,189 @@
+import org.apache.spark.sql.types.{StructField, StructType}
+
+/**
+ * Converts CSV string to a sequence of string
+ */
+private[csv] object UnivocityParser extends Logging {
+  /**
+   * Convert the input iterator to a iterator having [[InternalRow]]
+   */
+  def parseCsv(
+      iter: Iterator[String],
+      schema: StructType,
+      requiredSchema: StructType,
+      headers: Array[String],
+      shouldDropHeader: Boolean,
+      options: CSVOptions): Iterator[InternalRow] = {
+    if (shouldDropHeader) {
+      CSVUtils.dropHeaderLine(iter, options)
+    }
+    val csv = CSVUtils.filterCommentAndEmpty(iter, options)
+
+    val schemaFields = schema.fields
+    val requiredFields = requiredSchema.fields
+    val safeRequiredFields = if (options.dropMalformed) {
+      // If `dropMalformed` is enabled, then it needs to parse all the values
+      // so that we can decide which row is malformed.
+      requiredFields ++ schemaFields.filterNot(requiredFields.contains(_))
+    } else {
+      requiredFields
+    }
+    val safeRequiredIndices = new Array[Int](safeRequiredFields.length)
+    schemaFields.zipWithIndex.filter {
+      case (field, _) => safeRequiredFields.contains(field)
+    }.foreach {
+      case (field, index) => safeRequiredIndices(safeRequiredFields.indexOf(field)) = index
+    }
+    val requiredSize = requiredFields.length
+
+    tokenizeData(csv, options, headers).flatMap { tokens =>
+      if (options.dropMalformed && schemaFields.length != tokens.length) {
+        logWarning(s"Dropping malformed line: ${tokens.mkString(options.delimiter.toString)}")
+        None
+      } else if (options.failFast && schemaFields.length != tokens.length) {
+        throw new RuntimeException(s"Malformed line in FAILFAST mode: " +
+          s"${tokens.mkString(options.delimiter.toString)}")
+      } else {
+        val indexSafeTokens = if (options.permissive && schemaFields.length > tokens.length) {
+          tokens ++ new Array[String](schemaFields.length - tokens.length)
+        } else if (options.permissive && schemaFields.length < tokens.length) {
+          tokens.take(schemaFields.length)
+        } else {
+          tokens
+        }
+        try {
+          val row = convertTokens(
+            indexSafeTokens,
+            safeRequiredIndices,
+            schemaFields,
+            requiredSize,
+            options)
+          Some(row)
+        } catch {
+          case NonFatal(e) if options.dropMalformed =>
+            logWarning("Parse exception. " +
+              s"Dropping malformed line: ${tokens.mkString(options.delimiter.toString)}")
+            None
+        }
+      }
+    }
+  }
+
+  /**
+   * Convert the tokens to [[InternalRow]]
+   */
+  private def convertTokens(
+      tokens: Array[String],
+      requiredIndices: Array[Int],
+      schemaFields: Array[StructField],
+      requiredSize: Int,
+      options: CSVOptions): InternalRow = {
+    val row = new GenericMutableRow(requiredSize)
```

Comment:
Oh yes! I noticed this too. JSON data source will does as far as I remember. This might have to be definitely changed. I think I could do this in another PR.